src/VisualStudio/Core/Def/Implementation/IntellisenseControls/IntellisenseTextBoxViewModelFactory.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Projection;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Microsoft.VisualStudio.Utilities;
+using ImportingConstructorAttribute = System.Composition.ImportingConstructorAttribute;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.IntellisenseControls
+{
+    [Export(typeof(IntellisenseTextBoxViewModelFactory)), Shared]
+    internal class IntellisenseTextBoxViewModelFactory
+    {
+        private static readonly INITVIEW[] s_InitViews = new[] {
+            new INITVIEW()
+            {
+                fSelectionMargin = 0,
+                fWidgetMargin = 0,
+                fDragDropMove = 0,
+                IndentStyle = vsIndentStyle.vsIndentStyleNone
+            }
+        };
+
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly ITextEditorFactoryService _textEditorFactoryService;
+        private readonly OLE.Interop.IServiceProvider _serviceProvider;
+        private readonly IProjectionBufferFactoryService _projectionBufferFactoryService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public IntellisenseTextBoxViewModelFactory(
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            ITextEditorFactoryService textEditorFactoryService,
+            IContentTypeRegistryService contentTypeRegistryService,
+            IProjectionBufferFactoryService projectionBufferFactoryService,
+            SVsServiceProvider services)
+        {
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _textEditorFactoryService = textEditorFactoryService;
+            _serviceProvider = (OLE.Interop.IServiceProvider)services.GetService(typeof(OLE.Interop.IServiceProvider));
+            _projectionBufferFactoryService = projectionBufferFactoryService;
+        }
+
+        public async Task<IntellisenseTextBoxViewModel[]> CreateIntellisenseTextBoxViewModelsAsync(
+          Document document,
+          IContentType contentType,
+          string contentString,
```

Comment:
Is contentString just the document text or were other changes already made? I think it's the latter and I'm wondering if we're better off replacing "Document" with a Project + DocumentId, or just expecting the caller has passed in a Document with the mutated document text. It seems like it'd be a very easy mistake to make where somebody thinks the textual contents are the same in this function...

Reply:
I removed the content string and create it within CreateIntellisenseTextBoxViewModels. 

We need the document in some places before calling the current method. Passing Project + DocumentId would require searching for the document multiple times. Why do we need this?

---
In reply to: [367186498](https://github.com/dotnet/roslyn/pull/38748#discussion_r367186498) [](ancestors = 367186498)